### PR TITLE
[#8025] backend audit log

### LINKED ISSUE #8025: [Audit|Backend] Write calls to backend methods into audit log [8]
### Situation Description

#8029 provides a service to write audit events into an audit log. We need to make use of this to create log entries for all calls to backend methods.

We will use the interceptor pattern, similar as we are doing it with `PerformanceLoggingInterceptor`.

### Feature Description

Implement the logging of "patient" related calls.

- [x] Add a `kogBackendCall` method to the `AuditLogger` (provided by #8029)
- [x] Parameteres are: called Class, called Method, List of parameters, return value
- [x] The user either also needs to be parameter or provided in another way
- [x] Check if the audit-log is active

In the `logBackendCall` method we need to fill the `AuditEvent` (similar to #8029).

 ~~type, subtype: TODO~~ tracked in #8464 separately
- [x] `action`: See below
- [x] `recorded`: current timestamp
- [x] `agent`: the user (uuid and user name). 
  Coding is "humanuser" from http://terminology.hl7.org/CodeSystem/extra-security-role-type
  who.identified is the uuid
  name is the user name
- [x] `source.site`: URL of the server - see #8029
- [x] `source.type`: see #8029
- [x] `source.observer`: class name + method name, e.g. `de.symeda.sormas.api.caze.CaseFacade.getIndexList`
- [x]    For uuid based DTOs or lists of DTOs log the UUID, e.g. CaseFacade.saveCase(HGT23S-....)
- [x] Result parameters: Same as for call parameters. E.g. CaseFacade.getIndexList would have a list of UUIDs to be shown here
- [x]  May be necessary to exclude methods from being logged (e.g. read access to regions, districts, ...)
- [x] Identify the **action** (Create/Read/Update/Delete/Execute) related to the call. We can't distinguish create and update, so we will only use update.
- [x] For uuid based DTOs or lists of DTOs log the UUID, e.g. CaseFacade.saveCase(HGT23S-....)
- [x] create a helper method and a mapping that takes the called Method object and identifies the action based on the name.
  * Read: count, get, is, has, does, validate
  * Update: save, update, post, set, archive, dearchive
  * Delete: delete, merge
  * Execute: send
- [x] create a unit test based on ArchUnit that goes through all classes/interface in sormas-api that have the Remote annotation. For each of them check whether they have a method whose name does not fall into the list above. Provide the list of method names that are not covered here, so we can create a follow-up issue to deal with them. The unit test probably needs to be uncommented until then

### Additional Information
    
* Call parameters:

  * ~~Provide annotation that allows to include specific fields in the log. Ideally as part of the class, so that we don't have to go through all fields~~
  * ~~**_Discuss_**: What about objects like CaseCriteria?~~ Leave it as it is
  * ~~**_Discuss_**: What about UUID strings? Use an annotation? Could later be replaced (see #7086)~~ Just leave them

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #8025

### CODE REVIEW

**Comment by JonasCir** [comment]
@MartinWahnschaffe 
A couple of small todos left, but overall good for a fist pass :)

**Comment by JonasCir** [comment]
@MartinWahnschaffe This is as optimized as it gets for now. We have a dedicated fast path for this feature if disabled. When enabled, I did not witness any noticeable loss of performance.

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]

File: sormas-backend/src/main/java/de/symeda/sormas/backend/audit/AuditLogger.java [hunk 1]
```diff
@@ -15,59 +15,82 @@
...
-				logger.error("Could not fetch ConfigFacade for audit logger.");
-				throw new RuntimeException(e);
-			}
+	@PostConstruct
+	private void setup() {
+		String sourceSite = configFacade.getAuditSourceSite();
+		if (sourceSite.equals("")) {
+			logger.warn("audit.source.site is empty! Please configure it for more expedient audit trail analysis.");
+			sourceSite = "NOT CONFIGURED";
+			loggingDisabled = true;
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/audit/AuditLogger.java]: done

File: sormas-backend/src/main/java/de/symeda/sormas/backend/audit/AuditLoggerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+			// ignore previously discovered local beans
+			return context.proceed();
+		}
+
+		if (target.getAnnotationsByType(LocalBean.class).length > 0) {
+			// with this we ignore EJB calls which definitely originate from within the backend
+			// as they can never be called direct from outside (i.e., remote) of the backend
+
+			// cache that this class is local
+			ignoreLocalAnnotatedClasses.add(target);
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/audit/AuditLoggerInterceptor.java]: fixed

**Review by MartinWahnschaffe — APPROVED** [review]

**Pull request merged**